v2/pkg/core/execute.go

```diff
@@ -0,0 +1,131 @@
+package core
+
+import (
+	"fmt"
+
+	"github.com/projectdiscovery/gologger"
+	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/clusterer"
+	"github.com/projectdiscovery/nuclei/v2/pkg/templates"
+	"github.com/projectdiscovery/nuclei/v2/pkg/templates/types"
+	"github.com/remeh/sizedwaitgroup"
+	"github.com/rs/xid"
+	"go.uber.org/atomic"
+)
+
+// Execute takes a list of templates/workflows that have been compiled
+// and executes them based on provided concurrency options.
+//
+// All the execution logic for the templates/workflows happens in this part
+// of the engine.
+func (e *Engine) Execute(templates []*templates.Template, input InputProvider) *atomic.Bool {
+	return e.ExecuteWithOpts(templates, input, false)
+}
+
+// ExecuteWithOpts is execute with the full options
+func (e *Engine) ExecuteWithOpts(templatesList []*templates.Template, input InputProvider, noCluster bool) *atomic.Bool {
+	var finalTemplates []*templates.Template
+	if !noCluster {
+		finalTemplates, _ = e.ClusterTemplates(templatesList)
+	} else {
+		finalTemplates = templatesList
+	}
+
+	results := &atomic.Bool{}
+	for _, template := range finalTemplates {
+		templateType := template.Type()
+
+		var wg *sizedwaitgroup.SizedWaitGroup
+		if templateType == types.HeadlessProtocol {
+			wg = e.workPool.Headless
+		} else {
+			wg = e.workPool.Default
+		}
+
+		wg.Add()
+		switch {
+		case template.SelfContained:
+			// Self Contained requests are executed here separately
+			e.executeSelfContainedTemplateWithInput(template, results)
+		default:
+			// All other request types are executed here
+			e.executeModelWithInput(templateType, template, input, results)
+		}
+		wg.Done()
+	}
+	e.workPool.Wait()
+	return results
+}
+
+// processSelfContainedTemplates execute a self-contained template.
+func (e *Engine) executeSelfContainedTemplateWithInput(template *templates.Template, results *atomic.Bool) {
+	match, err := template.Executer.Execute("")
```

Comment:
You might consider adding a new `ExecuteSelfContained()` method that would delegate to the `Execute` method with the `""` parameter to make it understandable at first sight

Reply:
As not all request types implement it, I felt this was a cleaner way to do this without making it overly complex.